File: src/challenges/go/problems/max_length_between_equal_characters_test.go
// https://leetcode.com/problems/largest-substring-between-two-equal-characters

package problems

import (
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type MaxLengthBetweenEqualCharacters struct{}

func TestMaxLengthBetweenEqualCharacters(t *testing.T) {
	gen := core.TestSuite[MaxLengthBetweenEqualCharacters]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param("aa").Result(0)
	}).Add(func(tc *core.TestCase) {
		tc.Param("abca").Result(2)
	}).Add(func(tc *core.TestCase) {
		tc.Param("cbzxy").Result(-1)
	}).Run(t)
}

func (MaxLengthBetweenEqualCharacters) Solution(s string) int {
	set := map[byte]int{}

	result := -1
	for i := 0; i < len(s); i++ {
		if _, ok := set[s[i]]; ok {
			diff := i - set[s[i]] - 1
			if result < diff {
				result = diff
			}
		} else {
			set[s[i]] = i
		}
	}

	return result
}

File: src/challenges/python/problems/max_length_between_equal_characters_test.py
# https://leetcode.com/problems/largest-substring-between-two-equal-characters

from core.problem_base import *

class MaxLengthBetweenEqualCharacters(ProblemBase):
    def Solution(self, s: str) -> int:
        map = {}

        result = -1
        for i in range(len(s)):
            if s[i] in map:
                result = max(result, i - map[s[i]] - 1)
            else:
                map[s[i]] = i

        return result

if __name__ == '__main__':
    TestGen(MaxLengthBetweenEqualCharacters) \
        .Add(lambda tc: tc.Param("aa").Result(0)) \
        .Add(lambda tc: tc.Param("abca").Result(2)) \
        .Add(lambda tc: tc.Param("cbzxy").Result(-1)) \
        .Run()
